fixes:#18478 touch preventing mouseup in IE10+

nicknisi opened the pull request:
When touch is loaded on the page, it catches the `mouseup` event in
IE10/11. This is new in Dojo 1.10, as it appears it did not do this in version 1.9. See [example fiddle](http://jsfiddle.net/nicknisi/avr9eyok/). fixes:[#18478](https://bugs.dojotoolkit.org/ticket/18478). This change looks big, but it is mostly whitespace changes.

wkeese commented:
I'll look into the bug, but as for your patch, it doesn't make sense to me.   You have this code:

	if (hasTouch) {
		if(hasPointer){

ISTM that `hasTouch` and `hasPointer` will never be both true, and in particular on IE `hasTouch` will be false.   So IIUC your patch just disables the dojo/touch code completely on IE10+.   Arguably the code isn't very useful on IE, but probably we can't just disable it completely.

nicknisi commented:
Wouldn't `has('touch-events')` and `has('pointer-events')` both be true on a Windows Surface tablet running IE11? I created this assuming both could be true, and that touch wouldn't be necessary on a non-touch device. I can dig deeper in to this. Thanks for the feedback.

wkeese commented:
> Wouldn't has('touch-events') and has('pointer-events') both be true on a Windows Surface tablet running IE11? 

No, although has("touch") would be true.   has("touch-events") is specifically for touchstart/touchmove/touchend events.

nicknisi commented:
Thanks for pointing this out. I didn't realize the `hasTouch` variable had changed. I updated the PR to check `has("touch")` instead. I also dug deeper in to this to see if there was a way around it. The `mouseup` and `mousedown` events are being suppressed because they will occur significantly after the `click` event, which isn't good. What if we fire these events synthetically as well? This would resolve issues people have that rely on these events being fired. The big issue with this approach is that the events won't fire in the correct order. `touchend` will fire before `mousedown`, and `blur` will fire before `mousedown`, which isn't correct. Unfortunately, I'm not really sure what we can do. I looked at what other solutions are doing, and this fix is also what [fastclick](https://github.com/ftlabs/fastclick/pull/308) is proposing, but of course they have the same concern about event order.

I've updated the pull request with these changes and have tested several the widgets that would be impacted by this, as not many set `dojoClick`. Do you have any ideas on how we could work around this?

wkeese commented:
I suppose we could emit synthetic mousedown/mouseup events (like you did in your second commit).  I don't have any better ideas.   You said that it was problematic because "touchend will fire before mousedown"... I actually thought that's how it was supposed to work.

As for your first commit, I see it's just to skip all the code on IE10+ machines without touch support... I guess that makes sense too, since the code was already skipped on IE6-9.

nicknisi commented:
Yeah, I think that's actually right regarding `touchend` before `mousedown`. So, the biggest potential issue is that `blur` will fire before `mousedown`.

wkeese commented:
I agree that is weird; hopefully it won't break anyone.

wkeese commented:
There's another issue though.  On other browsers, the order of events are 'touch.press, mousedown, touch.release, mouseup, click'. With your change, it seems like mousedown will occur after touch.release.

Further, in my tests I'm not getting notification for touch.press and touch.release at all.

So, this PR needs to be tested and debugged.  I started a test at https://github.com/wkeese/dojo/commit/4b3f1b25d154f232c5306e9a5d6c4879417781ab.  Please debug from there, and add the test to the PR.  Unfortunately some other tests in that file are also failing, but presumably that's not your fault.

(See the README in tests/ for how to run the tests.   You should update dojo.intern.js to only run tests/functional/touch.)

nicknisi commented:
Ok, Thank you for the info. I'll test and debug from your commit. Thanks!

nicknisi commented:
Can you provide more information about touch.press and touch.release not working? They seem to be working for me in Chrome, Chrome in touch notebook emulation, and IE10/11.

Additionally, here is the order that I am seeing for the events (as observed in IE11) observed on the master branch and on this branch, with `dojoClick` and without it. The only difference is the inclusion of `mouseup` in this branch when `dojoClick: true`.

## on master

| dojoClick: true  | dojoClick: false |
|------------------|------------------|
| pointerdown      | pointerdown      |
| touch.press      | touch.press      |
| mousedown        | mousedown        |
| pointerup        | pointerup        |
| touch.release    | touch.release    |
| click            | mouseup          |
|                  | click            |

## on this branch

| dojoClick: true  | dojoClick: false |
|------------------|------------------|
| pointerdown      | pointerdown      |
| touch.press      | touch.press      |
| mousedown        | mousedown        |
| pointerup        | pointerup        |
| touch.release    | touch.release    |
| mouseup          | mouseup          |
| click            | click            |

wkeese commented:
I can try again tomorrow, but did you try the test on my nicknisi-18478-mouseup-ie branch?

nicknisi commented:
Yep. I never got any failing tests in tests/functional/touch.js. I have also merged the tests into this pull request.

jbest84 commented:
+1
We are seeing this exact issue with some drag and drop code. MouseUp doesn't fire and the code still thinks the user is dragging the node. This pull request fixes it.

wkeese commented:
@nicknisi, sorry for the late response.  The test is working now for me too on FF, Chrome, and IE10.   But presumably the reason it's working is that (on your branch) most of the dojo/touch code is disabled when `has("touch")` is false.

As you listed above, the test file expects mousedown to occur before touch.release.

But on iOS, the order of *native* events is: touchstart, touchend, mousedown, mouseup, click.   Likewise, in your PR you have this code that also emits the mousedown after touchend:

```js
setTimeout(function(){
	on.emit(target, "mousedown", mouseDownEvt);
	on.emit(target, "mouseup", mouseUpEvt);
	on.emit(target, "click", clickEvt);

	// refresh clickTime in case app-defined click handler took a long time to run
	clickTime = (new Date()).getTime();
}, 0);
```

In other words, I tried this manual test file on Chrome/mac and iOS Safari:

```html
<html>
	<head>
		<script src="./dojo.js" data-dojo-config="async:true"></script>
		<script>
			require(["dojo/on", "dojo/touch", "dojo/domReady!"], function(on, touch){
				on(div, touch.press, function(){ console.log("touch.press"); });
				on(div, touch.release, function(){ console.log("touch.release"); });
			});
		</script>
	</head>
	<body>
		<div id=div style="width: 200px; height: 200px; background: blue">click me</div>
		<script>
			div.addEventListener("touchstart", function(){ console.log("touchstart"); });
			div.addEventListener("touchend", function(){ console.log("touchend"); });
			div.addEventListener("mousedown", function(){ console.log("mousedown"); });
			div.addEventListener("mouseup", function(){ console.log("mouseup"); });
			div.addEventListener("click", function(){ console.log("click"); });
		</script>
	</body>
</html>
```

On Chrome, I get this result in the console:

```
mousedown
touch.press
mouseup
touch.release
click
```

But on my iPad, I get this result (when it's tethered to my mac so I can debug in Safari):

```
touchstart (touch.html, line 14)
[Log] touch.press (touch.html, line 6)
[Log] touchend (touch.html, line 15)
[Log] touch.release (touch.html, line 7)
[Log] mousedown (touch.html, line 16)
[Log] mouseup (touch.html, line 17)
[Log] click (touch.html, line 18)
```

I imagine if I clicked Chrome's checkbox to enable touch events that it would behave like Safari.

I'm not sure what to make of all that.  The order of events is varying depending on the environment, but at least we are getting a mousedown.   Still, it's not ideal, and the automated test would presumably fail if run against iOS or Android.   Thoughts?

nicknisi commented:
@wkeese sorry for the late response. So I put some more work into investigating this and here are the results I have for several different environments. I have made bold the changes from master for each.

## On my branch

### `dojoClick: true`
| Chrome         | Chrome with touch  | IE11          | iOS8 (Simulator) | iOS8 (iPhone 6 Plus) |
| :------------: | :----------------: | :-----------: | :-----------:    | :-----------:        |
| mousedown      | touch.press   	  | touch.press   | touch.press      | touch.press          |
| touch.press    | touchstart     	  | pointerdown   | touchstart       | touchstart           |
| mouseup        | touchend  		  | mousedown     | touchend         | touchend             |
| touch.release  | touch.release      | touch.release | touch.release    | touch.release        |
| click          | **mousedown**	  | pointerup     | **mousedown**    | **mousedown**        |
|                | **mouseup** 		  | mouseup       | **mouseup**      | **mouseup**          |
|                | click     		  | click         | click            | click                |

### `dojoClick: false`
| Chrome         | Chrome with touch  | IE11          | iOS8 (Simulator) | iOS8 (iPhone 6 Plus) |
| :------------: | :----------------: | :-----------: | :-----------:    | :-----------:        |
| mousedown      | touch.press   	  | touch.press   | touch.press      | touch.press          |
| touch.press    | touchstart     	  | pointerdown   | touchstart       | touchstart           |
| mouseup        | touchend  		  | mousedown     | touchend         | touchend             |
| touch.release  | touch.release      | touch.release | touch.release    | touch.release        |
| click          | mousedown    	  | pointerup     | mousedown        | mousedown            |
|                | mouseup   		  | mouseup       | mouseup          | mouseup              |
|                | click     		  | click         | click            | click                |

## On master

### `dojoClick: true`
| Chrome         | Chrome with touch  | IE11          | iOS8 (Simulator) | iOS8 (iPhone 6 Plus) |
| :------------: | :----------------: | :-----------: | :-----------:    | :-----------:        |
| mousedown      | touch.press   	  | touch.press   | touch.press      | touch.press          |
| touch.press    | touchstart   	  | pointerdown   | touchstart       | touchstart           |
| mouseup        | touchend  		  | mousedown     | touchend         | touchend             |
| touch.release  | touch.release      | touch.release | touch.release    | touch.release        |
| click          | click     		  | pointerup     | click            | click                |
|                |           		  | **mouseup**   |                  |                      |
|                |           		  | click         |                  |                      |

### `dojoClick: false`
| Chrome         | Chrome with touch  | IE11          | iOS8 (Simulator) | iOS8 (iPhone 6 Plus) |
| :------------: | :----------------: | :-----------: | :-----------:    | :-----------:        |
| mousedown      | touch.press        | touch.press   | touch.press      | touch.press          |
| touch.press    | touchstart 		  | pointerdown   | touchstart       | touchstart           |
| mouseup        | touchend  		  | mousedown     | touchend         | touchend             |
| touch.release  | touch.release      | touch.release | touch.release    | touch.release        |
| click          | mousedown 		  | pointerup     | mousedown        | mousedown            |
|                | mouseup   		  | mouseup       | mouseup          | mouseup              |
|                | click     		  | click         | click            | click                |


## On 98d78d464dfcce4411f04f9f398f8bfc8580f544 (`master~1`)

### `dojoClick: true`
| Chrome         | Chrome with touch  | IE11          | iOS8 (Simulator) | iOS8 (iPhone 6 Plus) |
| :------------: | :----------------: | :-----------: | :-----------:    | :-----------:        |
| mousedown      | touch.press   	  | touch.press   | touch.press      | touch.press          |
| touch.press    | touchstart   	  | pointerdown   | touchstart       | touchstart           |
| mouseup        | touchend  		  | mousedown     | touchend         | touchend             |
| touch.release  | touch.release      | touch.release | touch.release    | touch.release        |
| click          | click     		  | pointerup     | click            | click                |
|                |           		  | click         |                  |                      |
|                |           		  |               |                  |                      |

### `dojoClick: false`
| Chrome         | Chrome with touch  | IE11          | iOS8 (Simulator) | iOS8 (iPhone 6 Plus) |
| :------------: | :----------------: | :-----------: | :-----------:    | :-----------:        |
| mousedown      | touch.press        | touch.press   | touch.press      | touch.press          |
| touch.press    | touchstart 		  | pointerdown   | touchstart       | touchstart           |
| mouseup        | touchend  		  | mousedown     | touchend         | touchend             |
| touch.release  | touch.release      | touch.release | touch.release    | touch.release        |
| click          | mousedown 		  | pointerup     | mousedown        | mousedown            |
|                | mouseup   		  | mouseup       | mouseup          | mouseup              |
|                | click     		  | click         | click            | click                |

And here is the test code I was using:

```javascript
require([
	'dojo/on',
	'dojo/touch',
	'dojo/dom-construct',
	'dojo/domReady!'
], function(on, touch, domConstruct) {
	[
		domConstruct.create('input', {
			type: 'text',
			placeholder: 'dojoClick: false',
			style: 'border:1px solid green;margin:10px'
		}, document.body),
		domConstruct.create('input', {
			type: 'text',
			dojoClick: true,
			placeholder: 'dojoClick: true',
			style: 'border:1px solid red;margin:10px'
		}, document.body)
	].forEach(function (el) {
		[
			'click',
			'mouseup',
			'mousedown',
			{ name: 'touch.press', event: touch.press },
			{ name: 'touch.release', event: touch.release },
			'touchstart',
			'touchend',
			'pointerup',
			'pointerdown',
			'blur'
		].forEach(function (eventType) {
			var dojoClick = el.dojoClick;
			var event = eventType.event || eventType;
			var name = (eventType.name || eventType).toUpperCase();
			on(el, event, function (event) {
				console.log('%c' + name, 'font-weight: bold;color:' + (dojoClick ? 'red' : 'green') + ';');
			});
		});
	});
});
```

One thing I did notice is that f8c747b2399f88f9a3a1f116afc955d5ae39be0d which you merged onto master actually does resolve the issues of IE not emitting a `mouseup` event. I suppose with that, this PR could be closed, as I'm not really sure it's necessary for iOS to emit a `mousedown`/`mouseup` event, but it does make it more consistent with the non-`dojoClick` behavior. What do you think?

wkeese commented:
> One thing I did notice is that f8c747b which you merged onto master actually does resolve the issues of IE not emitting a mouseup event.

I think you mean IE desktop.  Presumably IE on Windows Surface still has the same problem.   Admittedly that's much less important that IE desktop not working...

> I suppose with that, this PR could be closed, as I'm not really sure it's necessary for iOS to emit a mousedown/mouseup event.

I don't see any meaningful distinction between iOS and IE.   Why is it important to have mouseup on IE, but not important on other browsers?   I realize that an app doesn't really *need* mouseup, because it instead can listen for touchend and pointerup, but still, if we fix it for one browser we might as well fix it for all browsers.

Again, my only concern was the order of events, specifically whether the mousedown occurred when you start touching or stop touching the screen, but I suppose it doesn't matter much, since the order is also inconsistent with vanilla web applications (i.e. when dojo/touch isn't involved).

So, I guess I can check this in basically as-is.

wkeese commented:
OK, I checked in your fix to emit mousedown/mouseup as 945c52c8c0cb59d80c06f424d3e21a754686ac9a, and a modified more lenient version of my test as b6721c0f2fa94ff383964e3a4ebca5e82b455784.   Thanks for the patches!

nicknisi commented:
Thank you!
